## Part 1 - Exploratory Data Analysis

### 1.0. Libraries and data files
***

In [1]:
!apt-get update
# importing the necessary spark environment and getting jdk
!apt-get install openjdk-8-jdk-headless -qq > /dev/null
!wget -q https://www-us.apache.org/dist/spark/spark-3.1.2/spark-3.1.2-bin-hadoop2.7.tgz
#!wget -q https://downloads.apache.org/spark/spark-3.1.2/spark-3.1.2-bin-hadoop2.7.tgz
!tar xf spark-3.1.2-bin-hadoop2.7.tgz
import os
os.environ['JAVA_HOME'] = '/usr/lib/jvm/java-8-openjdk-amd64'
os.environ['SPARK_HOME'] = '/content/spark-3.1.2-bin-hadoop2.7'

Get:1 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Ign:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
Get:3 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]
Get:4 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]
Ign:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Hit:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release
Hit:7 http://archive.ubuntu.com/ubuntu bionic InRelease
Hit:8 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Get:9 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Hit:10 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease
Get:11 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages [2,221 kB]
Hit:12 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease
Get:13

In [2]:
# getting the data
!pip install -q wget
!pip install -q findspark
import wget
data_19 = 'https://raw.githubusercontent.com/Gooner12/FIFA_2019/main/FIFA_19_players.csv'
data_21 = 'https://raw.githubusercontent.com/Gooner12/FIFA_2019/main/FIFA_21_players.csv'
dataset_19 = wget.download(data_19)
dataset_21 = wget.download(data_21)

In [3]:
import findspark
findspark.init()
#findspark.find()
from pyspark.sql import SparkSession

### 1.1 Data Exploration
***

In [4]:
# importing the 'FIFA_19.csv' as a Spark dataframe
spark = SparkSession.builder.appName('FIFA_19').config('spark.sql.caseSensitive', 'true').getOrCreate()
sc = spark.sparkContext
# not including header as data while reading so the header value is made true
df_19 = spark.read.format('csv').options(header='true', inferschema='true').load('FIFA_19_players.csv')
df_19.show()

+------+-----------------+---+--------------------+-----------+--------------------+-------+---------+-------------------+--------------------+--------+-------+-------+--------------+------------------------+---------+-----------+--------------+----------+---------+--------+-------------+------------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+
|    ID|             Name|Age|               Photo|Nationality|                Flag|Overall|Potential|               Club|           Club Logo|value(M)|wage(K)|Special|Preferred Foot|International Reputation|Weak Foot|Skill Moves|     Work Rate| Body 

In [5]:
df_21 = spark.read.format('csv').options(header='true', inferschema='true').load('FIFA_21_players.csv')
df_21.show()

+---------+--------------------+-----------------+--------------------+---+----------+---------+---------+-----------+-------------------+--------------------+-----------+-------+---------+---------+--------+----------------+--------------+------------------------+---------+-----------+-------------+--------------------+---------+------------------+--------------------+-------------+------------------+-----------+----------+--------------------+---------------+--------------------+----+--------+-------+---------+---------+------+---------+-----------+----------+-----------+--------+--------------+--------------------+------------------+-------------------+--------------------------+-----------------------+-----------------+---------------+-----------+-----------------+------------------+------------------+---------------------+---------------------+----------------+------------------+----------------+----------------+-------------+-------------+--------------+----------------+---------

Merging the 2019 and 2021 dataframe to get the overall, potential, value in 2021 for every player.

In [6]:
# first renaming the id column in fifa 21 database same as the id column in fifa 19 database for merging later
df_21 = df_21.withColumnRenamed('sofifa_id','ID')
# joining the two dataframes on id and only merging overall, potential and value_eur to the fifa 19 database
df = df_19.join(df_21['ID','overall','potential','value_eur'], on=['ID'], how='left_outer')
df.cache()
df.show()

+------+-----------------+---+--------------------+-----------+--------------------+-------+---------+-------------------+--------------------+--------+-------+-------+--------------+------------------------+---------+-----------+--------------+----------+---------+--------+-------------+------------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+-------+---------+---------+
|    ID|             Name|Age|               Photo|Nationality|                Flag|Overall|Potential|               Club|           Club Logo|value(M)|wage(K)|Special|Preferred Foot|International Reputation|Weak Foot|Skill

In [7]:
df.printSchema()

root
 |-- ID: integer (nullable = true)
 |-- Name: string (nullable = true)
 |-- Age: integer (nullable = true)
 |-- Photo: string (nullable = true)
 |-- Nationality: string (nullable = true)
 |-- Flag: string (nullable = true)
 |-- Overall: integer (nullable = true)
 |-- Potential: integer (nullable = true)
 |-- Club: string (nullable = true)
 |-- Club Logo: string (nullable = true)
 |-- value(M): string (nullable = true)
 |-- wage(K): integer (nullable = true)
 |-- Special: integer (nullable = true)
 |-- Preferred Foot: string (nullable = true)
 |-- International Reputation: integer (nullable = true)
 |-- Weak Foot: integer (nullable = true)
 |-- Skill Moves: integer (nullable = true)
 |-- Work Rate: string (nullable = true)
 |-- Body Type: string (nullable = true)
 |-- Real Face: string (nullable = true)
 |-- Position: string (nullable = true)
 |-- Jersey Number: integer (nullable = true)
 |-- Joined: string (nullable = true)
 |-- Loaned From: string (nullable = true)
 |-- Contract 

Dealing Missing Values:

In this stage, we examine the dataset to see the necessary operations required on data to ease our analysis. Based on the examination, we develop separate transformers that can perform data manipulation to output the data in a desired format.

In [8]:
# checking the missing values
from pyspark.sql.functions import count, col, when
from pyspark.sql import functions as F
null_values_count = df.select([count(when(F.isnull(cols), cols)).alias(cols) for cols in df.columns])
null_values_count.show()

+---+----+---+-----+-----------+----+-------+---------+----+---------+--------+-------+-------+--------------+------------------------+---------+-----------+---------+---------+---------+--------+-------------+------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+-------+---------+---------+
| ID|Name|Age|Photo|Nationality|Flag|Overall|Potential|Club|Club Logo|value(M)|wage(K)|Special|Preferred Foot|International Reputation|Weak Foot|Skill Moves|Work Rate|Body Type|Real Face|Position|Jersey Number|Joined|Loaned From|Contract Valid Until|Height|Height(CM)|Weight|Weight(KG)|Crossing|Finishing|Hea

In this analysis, the attributes that we are concerned with are player attributes that has impacts on players' gameplay such as 'Crossing', ShortPassing', etc. We are only concerned with forwards, midfielders and defenders not goalkeepers. It is seen that all attributes of interests have 48 missing values which include goalkeeping attributes too. So, we need to see if those missing values for different attributes belong to goalkeepers only before performing imputation or deletion of missing values.

In [9]:
# checking the missing values after removing goalkeepers
df_non_goalkeeper = df.filter(df['Position'] != 'GK')
df_non_goalkeeper.select([count(F.when(F.isnull(cols), cols)).alias(cols) for cols in df.columns]).show()

+---+----+---+-----+-----------+----+-------+---------+----+---------+--------+-------+-------+--------------+------------------------+---------+-----------+---------+---------+---------+--------+-------------+------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+-------+---------+---------+
| ID|Name|Age|Photo|Nationality|Flag|Overall|Potential|Club|Club Logo|value(M)|wage(K)|Special|Preferred Foot|International Reputation|Weak Foot|Skill Moves|Work Rate|Body Type|Real Face|Position|Jersey Number|Joined|Loaned From|Contract Valid Until|Height|Height(CM)|Weight|Weight(KG)|Crossing|Finishing|Hea

It is seen that no missing values are present in our many attributes of interests for players belonging to positions other than goalkeepers. Important features in our analysis such as value(M), potential and value_eur, still have missing values, so appropriate operations to deal with those missing values should be performed before we proceed to other analysis.

Handling of missing values in Value_2019(M) column:

In [10]:
from pyspark.sql import SQLContext
sqlContext = SQLContext(spark)

#df_view = 'soccer_analysis'
df.createOrReplaceTempView('soccer_analysis')

In [11]:
spark.sql("""
          Select * from soccer_analysis Where value_eur Is Not Null AND `value(M)` Is Null
          """).show()

+------+-------------+---+--------------------+------------+--------------------+-------+---------+---------------+--------------------+--------+-------+-------+--------------+------------------------+---------+-----------+--------------+---------+---------+--------+-------------+-----------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+-------+---------+---------+
|    ID|         Name|Age|               Photo| Nationality|                Flag|Overall|Potential|           Club|           Club Logo|value(M)|wage(K)|Special|Preferred Foot|International Reputation|Weak Foot|Skill Moves|     Work

It is seen that although we have missing values for players value in 2019, we can find the value of some of those players in 2021. Based on the value in 2021, we perform imputation of missing values in value column in 2019. To do that, we first compare the values in both years and calculate the percentage variation in values in those two years. Using that variation, we calculate the figures for missing values. 

The necessary operation is performed in a separate python file which we will import later.

In [12]:
spark.sql("""
          Select * from soccer_analysis Where `value(M)` Like '%K'
          """).show()

+------+-------------+---+--------------------+--------------+--------------------+-------+---------+--------------------+--------------------+--------+-------+-------+--------------+------------------------+---------+-----------+--------------+---------+---------+--------+-------------+------------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+-------+---------+---------+
|    ID|         Name|Age|               Photo|   Nationality|                Flag|Overall|Potential|                Club|           Club Logo|value(M)|wage(K)|Special|Preferred Foot|International Reputation|Weak Foot|Skill 

All the issues seen above have to be dealt before proceeding to further analysis. To deal with those issues, a separate python script has been developed which we will import later on in this analysis to create dataframes that are free from complications seen above.

Creating Pipelines for Data Preprocessing:

In this stage, we create a pipelines for the regression task. 

Note: the modules used to generate the regression dataframe can be found in the github repository for this project.

Here, we import two python scripts containing transformers that we developed according to our examination in the previous section. The first python script is data_preprocessor.py which contains classes such as ValueManipulator, UnifyValue and ValueImputer. The second script is target_generator, and it contains classes such as ProfitGenerator and GrowthGenerator.

In [13]:
!rm -rf FIFA_2019

In [14]:
# only for google colab
!git clone https://github.com/Gooner12/FIFA_2019.git FIFA_2019
!mv "/content/FIFA_2019/data_preprocessor.py" "/content/"
!mv "/content/FIFA_2019/target_generator.py" "/content/"
!mv "/content/FIFA_2019/encoder.py" "/content/"

Cloning into 'FIFA_2019'...
remote: Enumerating objects: 243, done.
remote: Counting objects: 100% (243/243), done.
remote: Compressing objects: 100% (152/152), done.
remote: Total 243 (delta 134), reused 168 (delta 90), pack-reused 0
Receiving objects: 100% (243/243), 4.95 MiB | 9.80 MiB/s, done.
Resolving deltas: 100% (134/134), done.


In [15]:
from pyspark.ml import Pipeline
import data_preprocessor, target_generator

# defining the stage that changes the units of players' values in 2021 along with renaming columns based on the year 
value_manipulation = data_preprocessor.ValueManipulator()
# defining the stage that standardises the players' values in 2019 to millions
value_standardisation = data_preprocessor.UnifyValue()
# defining the stage that imputes and removes missing values present in players' values in 2019 and 2021
value_imputation = data_preprocessor.ValueImputer()
# defining the stage that generates labels based on profits in two years for regression
regression_target = target_generator.ProfitGenerator()
# defining the stage that generates labels based on the potential to grow in two years
classification_target = target_generator.GrowthGenerator()

DataFrame for Regression:

In [16]:
# creating a pipeline to generate dataframe for regression
pipeline_regression = Pipeline(stages=[value_manipulation, value_standardisation, value_imputation, regression_target])
# fitting the pipeline model and transforming the data as defined
pipeline_model = pipeline_regression.fit(df)
regression_df = pipeline_model.transform(df)

In [17]:
regression_df.printSchema()

root
 |-- ID: integer (nullable = true)
 |-- Name: string (nullable = true)
 |-- Age: integer (nullable = true)
 |-- Photo: string (nullable = true)
 |-- Nationality: string (nullable = true)
 |-- Flag: string (nullable = true)
 |-- Overall_2019: integer (nullable = true)
 |-- Potential_2019: integer (nullable = true)
 |-- Club: string (nullable = true)
 |-- Club Logo: string (nullable = true)
 |-- Value_2019(M): double (nullable = true)
 |-- wage(K): integer (nullable = true)
 |-- Special: integer (nullable = true)
 |-- Preferred Foot: string (nullable = true)
 |-- International Reputation: integer (nullable = true)
 |-- Weak Foot: integer (nullable = true)
 |-- Skill Moves: integer (nullable = true)
 |-- Work Rate: string (nullable = true)
 |-- Body Type: string (nullable = true)
 |-- Real Face: string (nullable = true)
 |-- Position: string (nullable = true)
 |-- Jersey Number: integer (nullable = true)
 |-- Joined: string (nullable = true)
 |-- Loaned From: string (nullable = true)

In [18]:
regression_df.show(5)

+------+-----------------+---+--------------------+-----------+--------------------+------------+--------------+-------------------+--------------------+-------------+-------+-------+--------------+------------------------+---------+-----------+--------------+----------+---------+--------+-------------+------------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+------------+--------------+-------------+-----------------+
|    ID|             Name|Age|               Photo|Nationality|                Flag|Overall_2019|Potential_2019|               Club|           Club Logo|Value_2019(M)|wage(K)|S

In [19]:
# checking the missing values in regression dataframe
regression_df.select([count(F.when(F.isnull(cols), cols)).alias(cols) for cols in regression_df.columns]).show()

+---+----+---+-----+-----------+----+------------+--------------+----+---------+-------------+-------+-------+--------------+------------------------+---------+-----------+---------+---------+---------+--------+-------------+------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+------------+--------------+-------------+-----------------+
| ID|Name|Age|Photo|Nationality|Flag|Overall_2019|Potential_2019|Club|Club Logo|Value_2019(M)|wage(K)|Special|Preferred Foot|International Reputation|Weak Foot|Skill Moves|Work Rate|Body Type|Real Face|Position|Jersey Number|Joined|Loaned From|Contract Valid Un

It is seen that the column wage has 89 missing values, and this column might be important in predicting the players' price. 

In [20]:
# creating a view
regression_df.createOrReplaceTempView('soccer_analysis2')

# selecting the records containing null values in the wage column
spark.sql("""
          Select * from soccer_analysis2 Where `wage(K)` Is Null
          """).show()

+------+-------------+---+--------------------+------------+--------------------+------------+--------------+----+--------------------+-------------+-------+-------+--------------+------------------------+---------+-----------+--------------+---------+---------+--------+-------------+------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+------------+--------------+-------------+--------------------+
|    ID|         Name|Age|               Photo| Nationality|                Flag|Overall_2019|Potential_2019|Club|           Club Logo|Value_2019(M)|wage(K)|Special|Preferred Foot|International Repu

In [21]:
# finding the maximum value of a player whose wage is not disclosed
spark.sql("""
          Select MAX(`Value_2019(M)`), MIN(`Value_2019(M)`), MAX(Age), Min(Age) from soccer_analysis2 Where `wage(K)` Is Null
          """).show()

+------------------+------------------+--------+--------+
|max(Value_2019(M))|min(Value_2019(M))|max(Age)|min(Age)|
+------------------+------------------+--------+--------+
|            10.132|               0.0|      38|      18|
+------------------+------------------+--------+--------+



It is seen that the players whose wages are not disclosed are low value players ranging between 0 and 10 million. This might occur due to numerous reasons such as some players are just starting their professional football, some are at the end of their career or some are not good enough to negotiate their wage. These all reasons might led the players to accept the price their club has offered which is usually minimum wage. Thus, we impute the missing wage with the minimum wage amount.

In [22]:
# selecting the minimum wage
spark.sql("""
          Select MIN(`wage(K)`) from soccer_analysis2 Where `wage(K)` Is Not Null
          """).show()

+------------+
|min(wage(K))|
+------------+
|           1|
+------------+



In [23]:
# imputing the missing wage values with one
regression_df = regression_df.fillna(1, subset=['wage(K)'])

In [24]:
# checking the missing values in regression dataframe
regression_df.select([count(F.when(F.isnull(cols), cols)).alias(cols) for cols in regression_df.columns]).show()

+---+----+---+-----+-----------+----+------------+--------------+----+---------+-------------+-------+-------+--------------+------------------------+---------+-----------+---------+---------+---------+--------+-------------+------+-----------+--------------------+------+----------+------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+--------+----------+---------+-------------+----------+-----------------+------------+--------------+-------------+-----------------+
| ID|Name|Age|Photo|Nationality|Flag|Overall_2019|Potential_2019|Club|Club Logo|Value_2019(M)|wage(K)|Special|Preferred Foot|International Reputation|Weak Foot|Skill Moves|Work Rate|Body Type|Real Face|Position|Jersey Number|Joined|Loaned From|Contract Valid Un

In [25]:
regression_df.select(['Value_2019(M)', 'Value_2021(M)']).show(5)

+-------------+-------------+
|Value_2019(M)|Value_2021(M)|
+-------------+-------------+
|        110.5|         67.5|
|         77.0|         46.0|
|        118.5|         90.0|
|        102.0|         87.0|
|         93.0|         58.0|
+-------------+-------------+
only showing top 5 rows



### 3.1 Data Preparation

We try to predict the value of players in two years using the attributes that might impact players' price tags and values.
***

Function:

Feature Engineering:

In this step, we try to find the features that are most correlated with the target (Gain in Two Years). From this step, we will only pick up those attributes who have at least of 10% correlation with the target whether negative or positive. We use both spearman and pearson correlation to determine this. Any attribute that pass this criterion on any of those two tests will be accepted into the next step.

In [27]:
from pyspark.ml.feature import VectorAssembler

features_col = [ 'Value_2021(M)', 'Age', 'Value_2019(M)', 'wage(K)', 'International Reputation', 'Height(CM)', 'Weight(KG)', 
                      'Crossing', 'Finishing', 'HeadingAccuracy', 
                      'ShortPassing', 'Volleys', 'Dribbling', 'Curve',
                      'FKAccuracy', 'LongPassing', 'BallControl', 
                      'Acceleration', 'SprintSpeed', 'Agility', 
                      'Reactions', 'Balance', 'ShotPower', 'Jumping', 
                      'Stamina', 'Strength', 'LongShots', 'Aggression', 
                      'Interceptions', 'Positioning', 'Vision', 'Penalties', 
                      'Composure', 'Marking', 'StandingTackle', 'SlidingTackle', 'Overall_2019']

vec_assembler = VectorAssembler(inputCols=features_col, outputCol='features')
df_class = vec_assembler.transform(regression_df).select('features')
df_class.show(5)

+--------------------+
|            features|
+--------------------+
|[67.5,31.0,110.5,...|
|[46.0,33.0,77.0,4...|
|[90.0,26.0,118.5,...|
|[87.0,27.0,102.0,...|
|[58.0,27.0,93.0,3...|
+--------------------+
only showing top 5 rows



In [39]:
regression_df[features_col].show(5)

+-------------+---+-------------+-------+------------------------+----------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----------+------------+-----------+-------+---------+-------+---------+-------+-------+--------+---------+----------+-------------+-----------+------+---------+---------+-------+--------------+-------------+------------+
|Value_2021(M)|Age|Value_2019(M)|wage(K)|International Reputation|Height(CM)|Weight(KG)|Crossing|Finishing|HeadingAccuracy|ShortPassing|Volleys|Dribbling|Curve|FKAccuracy|LongPassing|BallControl|Acceleration|SprintSpeed|Agility|Reactions|Balance|ShotPower|Jumping|Stamina|Strength|LongShots|Aggression|Interceptions|Positioning|Vision|Penalties|Composure|Marking|StandingTackle|SlidingTackle|Overall_2019|
+-------------+---+-------------+-------+------------------------+----------+----------+--------+---------+---------------+------------+-------+---------+-----+----------+-----------+-----

In [28]:
from pyspark.ml.stat import Correlation
#r1 = Correlation.corr(df_class, 'features', method='pearson').head()
pearson_corr = Correlation.corr(df_class, 'features', 'pearson').collect()[0][0]
print(str(pearson_corr))
#print('Pearson correlation matrix:\n' + str(r1[0]))

DenseMatrix([[ 1.        , -0.02770739,  0.85782914, ...,  0.08731344,
               0.06946364,  0.56321327],
             [-0.02770739,  1.        ,  0.08332409, ...,  0.20606472,
               0.1774763 ,  0.45704163],
             [ 0.85782914,  0.08332409,  1.        , ...,  0.09897944,
               0.07406407,  0.63777141],
             ...,
             [ 0.08731344,  0.20606472,  0.09897944, ...,  1.        ,
               0.96660681,  0.26505528],
             [ 0.06946364,  0.1774763 ,  0.07406407, ...,  0.96660681,
               1.        ,  0.22502964],
             [ 0.56321327,  0.45704163,  0.63777141, ...,  0.26505528,
               0.22502964,  1.        ]])


In [52]:
# alternative way to calculate pearson correlation
import six
for i in regression_df[features_col].columns:
  # checking if the instance is of string type and using six to write compatible code for both python2.x and 3.x
  if not(isinstance(regression_df[features_col].select(i).take(1)[0][0], six.string_types)):
    print('Correlation to Value_2021(M) for ', i, regression_df[features_col].stat.corr('Value_2021(M)',i, 'pearson'))

Correlation to Value_2021(M) for  Value_2021(M) 1.0
Correlation to Value_2021(M) for  Age -0.027707394939826886
Correlation to Value_2021(M) for  Value_2019(M) 0.8578291413547301
Correlation to Value_2021(M) for  wage(K) 0.6871367407656495
Correlation to Value_2021(M) for  International Reputation 0.5061458600475491
Correlation to Value_2021(M) for  Height(CM) 0.0164670638967895
Correlation to Value_2021(M) for  Weight(KG) 0.03158107549541299
Correlation to Value_2021(M) for  Crossing 0.2866238438200077
Correlation to Value_2021(M) for  Finishing 0.2605724384026795
Correlation to Value_2021(M) for  HeadingAccuracy 0.18726361492016722
Correlation to Value_2021(M) for  ShortPassing 0.4184740045460058
Correlation to Value_2021(M) for  Volleys 0.28333113158540485
Correlation to Value_2021(M) for  Dribbling 0.36555716346075073
Correlation to Value_2021(M) for  Curve 0.30461908642180074
Correlation to Value_2021(M) for  FKAccuracy 0.2538595086700156
Correlation to Value_2021(M) for  LongPass

In [29]:
# creating a datafame from the dense matrix
import pandas as pd

labels = features_col 
#creating a dataframe from the dense matrix
pearson_df = pd.DataFrame(pearson_corr.toArray(), index = [i for i in labels], columns = [i for i in labels])
pearson_df

Value_2021(M)       Age  ...  SlidingTackle  Overall_2019
Value_2021(M)                  1.000000 -0.027707  ...       0.069464      0.563213
Age                           -0.027707  1.000000  ...       0.177476      0.457042
Value_2019(M)                  0.857829  0.083324  ...       0.074064      0.637771
wage(K)                        0.687137  0.147143  ...       0.105531      0.577682
International Reputation       0.506146  0.257786  ...       0.089109      0.503091
Height(CM)                     0.016467  0.034791  ...       0.087486      0.060618
Weight(KG)                     0.031581  0.216236  ...       0.156786      0.185425
Crossing                       0.286624  0.264051  ...       0.079624      0.496591
Finishing                      0.260572  0.145055  ...      -0.585836      0.372765
HeadingAccuracy                0.187264  0.344698  ...       0.234819      0.467578
ShortPassing                   0.418474  0.307773  ...       0.204886      0.723317
Volleys                        0.283331  0.244909  ...      -0.429767      0.451528
Dribbling                      0.365557  0.105670  ...      -0.257165      0.516803
Curve                          0.304619  0.259670  ...      -0.145965      0.502946
FKAccuracy                     0.253860  0.306941  ...      -0.079740      0.456142
LongPassing                    0.311748  0.306181  ...       0.361529      0.585123
BallControl                    0.443295  0.252234  ...      -0.094252      0.718558
Acceleration                   0.201222 -0.205127  ...      -0.276443      0.150386
SprintSpeed                    0.206534 -0.197892  ...      -0.259669      0.169555
Agility                        0.197761 -0.014238  ...      -0.267220      0.244360
Reactions                      0.484908  0.456704  ...       0.213878      0.848466
Balance                        0.113784 -0.091420  ...      -0.181544      0.059031
ShotPower                      0.305394  0.305241  ...      -0.221512      0.564037
Jumping                        0.075775  0.187817  ...       0.181856      0.227846
Stamina                        0.237617  0.224108  ...       0.283203      0.461374
Strength                       0.078977  0.359905  ...       0.290333      0.341815
LongShots                      0.287624  0.272781  ...      -0.307065      0.502239
Aggression                     0.154577  0.397279  ...       0.606660      0.453645
Interceptions                  0.111360  0.280574  ...       0.908322      0.334237
Positioning                    0.294273  0.189634  ...      -0.408797      0.438635
Vision                         0.340133  0.239894  ...      -0.140969      0.524594
Penalties                      0.224936  0.249726  ...      -0.429067      0.391049
Composure                      0.430534  0.469837  ...       0.134986      0.802987
Marking                        0.110445  0.233899  ...       0.859465      0.307226
StandingTackle                 0.087313  0.206065  ...       0.966607      0.265055
SlidingTackle                  0.069464  0.177476  ...       1.000000      0.225030
Overall_2019                   0.563213  0.457042  ...       0.225030      1.000000

[37 rows x 37 columns]

Attributes that fail to satisfy the criterion from the pearson test are height(cm), weight(kg), crossing, dribbling, acceleration, sprint speed, agility, balance, jumping, stamina, strength, interceptions, marking, standing tackle and sliding tackle.

In [45]:
# creating a dataframe based on the spearman correlation
spearman_corr = Correlation.corr(df_class, 'features', 'spearman').collect()[0][0]
spearman_df = pd.DataFrame(spearman_corr.toArray(), index = [i for i in labels], columns = [i for i in labels])
spearman_df

Value_2021(M)       Age  ...  SlidingTackle  Overall_2019
Value_2021(M)                  1.000000 -0.031378  ...       0.161119      0.736222
Age                           -0.031378  1.000000  ...       0.219029      0.480980
Value_2019(M)                  0.852911  0.231094  ...       0.209610      0.936872
wage(K)                        0.616682  0.334480  ...       0.212508      0.793917
International Reputation       0.320234  0.268063  ...       0.126916      0.439964
Height(CM)                     0.018145  0.055577  ...       0.136737      0.067064
Weight(KG)                     0.042882  0.223385  ...       0.190744      0.176246
Crossing                       0.391831  0.290206  ...       0.107060      0.515703
Finishing                      0.336049  0.156343  ...      -0.567660      0.374860
HeadingAccuracy                0.264642  0.357558  ...       0.298971      0.463664
ShortPassing                   0.573712  0.310933  ...       0.246005      0.721346
Volleys                        0.339630  0.252197  ...      -0.403681      0.434472
Dribbling                      0.544638  0.140607  ...      -0.213058      0.571272
Curve                          0.383999  0.271619  ...      -0.124868      0.499081
FKAccuracy                     0.305029  0.303860  ...      -0.084044      0.438080
LongPassing                    0.425783  0.312917  ...       0.367313      0.587677
BallControl                    0.630050  0.265100  ...      -0.040258      0.733255
Acceleration                   0.283449 -0.127902  ...      -0.264647      0.175200
SprintSpeed                    0.300918 -0.112156  ...      -0.236343      0.197933
Agility                        0.278653  0.028833  ...      -0.255118      0.262173
Reactions                      0.594043  0.479882  ...       0.270004      0.843757
Balance                        0.134347 -0.072572  ...      -0.189770      0.074134
ShotPower                      0.436190  0.339442  ...      -0.165429      0.593554
Jumping                        0.115608  0.232837  ...       0.223251      0.254075
Stamina                        0.323502  0.291573  ...       0.309671      0.469085
Strength                       0.120975  0.377519  ...       0.328636      0.343438
LongShots                      0.400749  0.303410  ...      -0.262089      0.529485
Aggression                     0.221337  0.415708  ...       0.612635      0.469947
Interceptions                  0.201256  0.312445  ...       0.901540      0.400234
Positioning                    0.410170  0.239529  ...      -0.371979      0.498886
Vision                         0.423714  0.259726  ...      -0.111738      0.539760
Penalties                      0.286102  0.246365  ...      -0.400564      0.379153
Composure                      0.537944  0.475939  ...       0.196319      0.793538
Marking                        0.181707  0.264212  ...       0.868778      0.357647
StandingTackle                 0.189402  0.245920  ...       0.960506      0.350047
SlidingTackle                  0.161119  0.219029  ...       1.000000      0.304961
Overall_2019                   0.736222  0.480980  ...       0.304961      1.000000

[37 rows x 37 columns]

Attributes that fail to satisfy the criterion from the spearman test are height(cm), acceleration, sprint speed, agility, balance and sliding tackle.

Based on the above two tests, we remove height(cm), acceleration, sprint speed, agility, balance and sliding tackle from the prediction analysis in further steps.

In [53]:
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.feature import StandardScaler

def steps_processor(features_col, df):

  # creating a feature vector
  vec_assembler = VectorAssembler(inputCols=features_col, outputCol='features')
  df_class = vec_assembler.transform(df).select('features','Value_2021(M)')

  # scaling the feature vector
  standardscaler = StandardScaler().setInputCol('features').setOutputCol('Scaled_features')
  raw_data = standardscaler.fit(df_class).transform(df_class)

  # splitting the dataset
  train, test = raw_data.randomSplit([0.7, 0.3], seed=2020)
  return train, test

In [54]:
features_col = ['Age', 'Value_2019(M)', 'wage(K)', 'International Reputation', 'Weight(KG)', 
                      'Crossing', 'Finishing', 'HeadingAccuracy', 
                      'ShortPassing', 'Volleys', 'Dribbling', 'Curve',
                      'FKAccuracy', 'LongPassing', 'BallControl', 
                      'Reactions', 'ShotPower', 'Jumping', 
                      'Stamina', 'Strength', 'LongShots', 'Aggression', 
                      'Interceptions', 'Positioning', 'Vision', 'Penalties', 
                      'Composure', 'Marking', 'StandingTackle', 'Overall_2019']
train, test = steps_processor(features_col, regression_df)

In [55]:
test.show()

+--------------------+-------------+--------------------+
|            features|Value_2021(M)|     Scaled_features|
+--------------------+-------------+--------------------+
|[17.0,1.100000023...|         0.95|[3.74211125958050...|
|[17.0,1.700000047...|        0.925|[3.74211125958050...|
|[17.0,1.799999952...|         45.0|[3.74211125958050...|
|[17.0,3.5,8.0,1.0...|          9.0|[3.74211125958050...|
|[17.0,3.5,25.0,1....|         10.0|[3.74211125958050...|
|[17.0,17.5,66.0,1...|         27.5|[3.74211125958050...|
|[18.0,1.0,2.0,1.0...|          1.6|[3.96223545132053...|
|[18.0,1.100000023...|          3.9|[3.96223545132053...|
|[18.0,1.100000023...|          1.1|[3.96223545132053...|
|[18.0,1.200000047...|         12.5|[3.96223545132053...|
|[18.0,1.200000047...|         0.85|[3.96223545132053...|
|[18.0,1.799999952...|          1.4|[3.96223545132053...|
|[18.0,2.200000047...|          2.0|[3.96223545132053...|
|[18.0,2.200000047...|         10.0|[3.96223545132053...|
|[18.0,2.59999

Next, we compare three baseline models and select the best baseline model for further tuning in the later step.

In [57]:
from pyspark.ml.regression import LinearRegression

lr = LinearRegression(featuresCol='Scaled_features', labelCol='Value_2021(M)', maxIter=10, regParam=0.3, elasticNetParam=0.8)
model_lr = lr.fit(train)
predict_train_lr = model_lr.transform(train)
print("Coefficients: " + str(model_lr.coefficients))
print("Intercept: " + str(model_lr.intercept))

Coefficients: [-0.4401235995808695,4.007756623677224,0.0,0.0,0.0,0.033373348502540263,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0656084537855371,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.013417635734974027,0.0,0.0,0.0,0.0,0.3900691234682971]
Intercept: -1.287181956175867


In [58]:
training_summary = model_lr.summary
print("RMSE: %f" % training_summary.rootMeanSquaredError)
print("r2: %f" % training_summary.r2)

RMSE: 2.644242
r2: 0.745382


In [60]:
# finding the mean, min and max to compare RMSE and measure performance
train.describe().show()

+-------+------------------+
|summary|     Value_2021(M)|
+-------+------------------+
|  count|             11204|
|   mean|2.5537435737236707|
| stddev|5.2405381631664785|
|    min|               0.0|
|    max|             105.5|
+-------+------------------+



R square value of 0.74, which is 74% variability in players' value in 2021 is explained using the above attributes and remaining 16% variability could be defined through other attributes seems good. And our RMSE of 2.64 considering the min, max and mean is not that bad. However, this result is for training set, we still need to see how our model performs for the test set.

In [66]:
from pyspark.ml.evaluation import RegressionEvaluator

predict_test_lr = model_lr.transform(test)
predict_test_lr.select('prediction','Value_2021(M)','features').show(5)
lr_evaluator = RegressionEvaluator(predictionCol='prediction', labelCol='Value_2021(M)', metricName='r2')
print('R Squared (R2) on test data = %g' % lr_evaluator.evaluate(predict_test_lr))

+------------------+-------------+--------------------+
|        prediction|Value_2021(M)|            features|
+------------------+-------------+--------------------+
| 2.285327161242521|         0.95|[17.0,1.100000023...|
|2.7436997653609883|        0.925|[17.0,1.700000047...|
|  2.86365007626162|         45.0|[17.0,1.799999952...|
| 4.206670232060807|          9.0|[17.0,3.5,8.0,1.0...|
| 4.204508393725057|         10.0|[17.0,3.5,25.0,1....|
+------------------+-------------+--------------------+
only showing top 5 rows

R Squared (R2) on test data = 0.750868


In [67]:
test_result = model_lr.evaluate(test)
print('Root Mean Squared Error (RMSE) on test data = %g' % test_result.rootMeanSquaredError)

Root Mean Squared Error (RMSE) on test data = 2.71582
